crates/cubecl-reduce/src/launch.rs

```diff
@@ -1,443 +1,159 @@
-        offset,
-        end,
-        stride,
-        accumulator_size,
-        line_size,
-        LineMode::Perpendicular,
-    );
-
-    sync_units();
-
-    let out = reduce_tree::<In, R>(&mut accumulator);
-    if UNIT_POS == 0 {
-        let out = R::to_output_perpendicular(out, input.shape(axis_reduce));
-
-        #[unroll]
-        for k in 0..line_size {
-            output[line_size * CUBE_POS + k] = out[k];
+    let accumulator = match comptime!((params.shared, params.use_planes)) {
+        (Some(accumulator_size), use_planes) => {
+            let mut accumulator = reduce_slice_shared::<In, R>(
+                input.to_slice(),
+                range,
+                accumulator_size,
+                params.line_size,
+                params.line_mode,
+                use_planes,
+            );
+            sync_units();
+            reduce_tree::<In, R>(&mut accumulator, accumulator_size)
         }
+        (None, true) => {
+            reduce_slice_plane::<In, R>(input.to_slice(), range, params.line_size, params.line_mode)
+        }
+        (None, false) => {
+            reduce_slice::<In, R>(input.to_slice(), range, params.line_size, params.line_mode)
+        }
+    };
+
+    if elected_writer(params) {
+        write_to_output::<In, Out, R>(
+            output,
+            accumulator,
+            reduce_index,
+            input.shape(axis_reduce),
+            params,
+        );
     }
 }
 
 #[cube]
-pub fn compute_reduce_offset<In: CubeType, Out: CubeType>(
-    index: u32,
-    input: &Tensor<In>,
-    output: &Tensor<Out>,
-    #[comptime] line_size: u32,
-) -> u32 {
-    let mut offset = 0;
-    for axis in 0..input.rank() {
-        let coordinate = output.coordinate(index, axis);
-        offset += coordinate * input.stride(axis);
-    }
-    offset / line_size
-}
-
-#[cube]
-pub fn reduce_slice<N: Numeric, R: Reduce<N>>(
-    items: Slice<Line<N>>,
-    start: u32,
-    end: u32,
-    stride: u32,
-    #[comptime] line_size: u32,
-    #[comptime] line_mode: LineMode,
-) -> R::AccumulatorItem {
-    let mut accumulator = R::null_accumulator(line_size);
-
-    let mut index = start;
-    let mut coordinate = 0;
-    while index < end {
-        let coordinates = match comptime!(line_mode) {
-            LineMode::Parallel => {
-                let mut coordinates = Line::empty(line_size).fill(coordinate * line_size);
-                if line_size > 1 {
-                    #[unroll]
-                    for j in 0..line_size {
-                        coordinates[j] += j;
-                    }
-                }
-                coordinates
-            }
-            LineMode::Perpendicular => Line::empty(line_size).fill(coordinate),
-        };
-        reduce_inplace::<N, R>(&mut accumulator, items[index], coordinates, false);
-        index += stride;
-        coordinate += 1;
+fn get_reduce_index(#[comptime] params: ReduceParams) -> u32 {
+    if params.shared.is_some() {
```

Comment:
The action depending on whether using shared memory, planes, or only units comes up frequently. It might be helpful to define an enum to represent these three strategies.